### PR TITLE
Fix errors in GUI WebSockets

### LINKED ISSUE #10381: Local projects leak PM connections
Opening multiple local projects results in periodically reconnects with previous ws connectors. 

1. Open a local project. 
2. Close it
3. Open another one. In console you will get `WebSocket connection to 'ws://127.0.0.1:59040/' failed:` to previous project url
4. Close and open another one. Now you will have 2 intervals trying to reconnect with previous urls

https://github.com/enso-org/cloud-v2/assets/2855109/4727cd7b-9ca9-471f-af80-f440d7e4c1af

### PR DESCRIPTION
**Pull request by somebody1234**
### Pull Request Description
- Fix https://github.com/enso-org/cloud-v2/issues/1337
- Fix #10381 (same issue)
  - This was caused by queued execution context tasks delaying the disposal of the WebSocket connections - I guess because the WebSocket is closed, either the queue never reaches that point, or retries take too long to reach that point?
  - There was also another issue caused by the Dashboard incorrectly caching the `project`, causing the endpoint URLs to become outdated.

Misc changes:
- Change `Editor.tsx` to not return two different component trees. This avoids React unmounting and re-mounting the GUI app.

### Important Notes
None

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] ~~The documentation has been updated, if necessary.~~
- [x] ~~Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.~~
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [x] ~~Unit tests have been written where possible.~~

### CODE REVIEW

**Comment by somebody1234** [comment]
pinging @MrFlashAccount for CR on the dashboard code changes. some of it may be out of place, I'd be happy to extract some of it to another PR

**Comment by MrFlashAccount** [comment]
@somebody1234  Code changes around editor/ProjectIcon will be completely overwritten by the PR where I'm fixing the project opening. Can I ask you to describe these changes so I can check that the bugs you fixed are also fixed in my pr?

**Review by MrFlashAccount — APPROVED** [review]
Approving changes in dashboard

**Comment by somebody1234** [comment]
ok:
- the `ProjectIcon` change was for debugging and can be safely ignored
- for `Editor` changes:
  - the change in the outer `Editor` component is the one that avoids unmounting + remounting the inner component
  - the change from `projectStartupInfo.projectAsset.id`  to `project.projectId` should not affect things after the `Suspense` was fixed, but they were previously out of sync due to the cache returning a value for `project` that was *different* to the current value of `projectStartupInfo.project` (which was a pending promise)
  - the `Promise.resolve` in the `queryFn` seems to fix React Suspense for some reason. for some reason it's broken when it returns a promise that is already fulfilled - perhaps this is due to caching?
  - the `queryKey` change is an oopsie, of course the `id` should be there... i think it made things marginally better, but i *think* in the current code the `Promise.resolve` is doing all the heavy lifting here? so this change might not be necessary
  - and finally, i dunno if the `gcTime` does anything relevant here but i added it just to be safe


tl;dr: the changes that matter the most are:
- the change in the outer `Editor` component
- the `Promise.resolve`
- (potentially) the `gcTime`

File: app/gui2/src/stores/project/executionContext.ts [hunk 1]
```diff
@@ -274,17 +274,22 @@ export class ExecutionContext extends ObservableV2<ExecutionContextNotification>
   }
 
   dispose() {
+    const abortController = new AbortController()
+    this.lsRpc.willRelease(abortController)
```
Comment (farmaazon) [inline app/gui2/src/stores/project/executionContext.ts]: `willRelease` is very misleading: I would expect something like "will release when aborted", but this actually aborts on _disconnect_

Also, I think `abortController` is not needed here at all; we just release LS once connection is closed.

File: app/gui2/src/stores/project/index.ts [hunk 2]
```diff
@@ -86,7 +86,6 @@ function createLsRpcConnection(clientId: Uuid, url: string, abort: AbortScope):
 function initializeDataConnection(clientId: Uuid, url: string, abort: AbortScope) {
   const client = createDataWebsocket(url, 'arraybuffer')
   const connection = new DataServer(clientId, client, abort)
-  abort.handleDispose(connection)
```
Comment (farmaazon) [inline app/gui2/src/stores/project/index.ts:89]: Why is this line removed?

File: app/gui2/src/stores/project/executionContext.ts (hunk 1 above)
Comment (farmaazon) [reply app/gui2/src/stores/project/executionContext.ts]: Also, I think it should be handled on higher level: in the future we might have more "cleanups" in LsRpc. So instead `willRelease` I would make method `stopReconnecting` and call it not here, but somewhere "higher" at the moment when we're sure the GUI app is tearing down.

**Review by farmaazon — CHANGES_REQUESTED** [review]

File: app/gui2/src/stores/project/index.ts (hunk 2 above)
Comment (somebody1234) [reply app/gui2/src/stores/project/index.ts:89]: i think DataServer already calls the exact same method on the same AbortScope, right?

File: app/gui2/src/stores/project/executionContext.ts (hunk 1 above)
Comment (somebody1234) [reply app/gui2/src/stores/project/executionContext.ts]: that sounds like it would make `retainCount`/`release`/`retain` redundant, right? since now `stopReconnecting` is what determines whether the LS transport/connection is disposed

Comment (farmaazon) [reply app/gui2/src/stores/project/executionContext.ts]: No, because you may close the GUI App without closing LS - in that case LS never closes the connection. I think this is still a case in cloud environment?

File: app/gui2/src/stores/project/index.ts (hunk 2 above)
Comment (farmaazon) [reply app/gui2/src/stores/project/index.ts:89]: Right.

**Comment by somebody1234** [comment]
dunno whether the new changes are correct

File: app/gui2/shared/languageServer.ts [hunk 3]
```diff
@@ -152,8 +153,13 @@ export class LanguageServer extends ObservableV2<Notifications & TransportEvents
     this.client.onError((error) => {
       console.error('Unexpected Language Server connection error:', error)
     })
-    transport.on('error', (error) => console.error('Language Server transport error:', error))
+    transport.on('error', (error) => {
+      if (this.shouldReconnect) {
+        console.error('Language Server transport error:', error)
+      }
```
Comment (farmaazon) [inline app/gui2/shared/languageServer.ts:159]: I think we should always print this? Error is an error

File: app/gui2/shared/languageServer.ts [hunk 4]
```diff
@@ -152,8 +153,13 @@ export class LanguageServer extends ObservableV2<Notifications & TransportEvents
...
+    })
     const reinitializeCb = () => {
+      if (!this.shouldReconnect) return
```
Comment (farmaazon) [inline app/gui2/shared/languageServer.ts]: The reconnecting mechanism is inside websocket inside `transport` - we need to call `close` if we don't want to reconnect anymore.

Comment (somebody1234) [reply app/gui2/shared/languageServer.ts]: i guess the issue (i think) is that `onAbort` seems to run quite a bit later than `stopReconnecting`, so without this condition then it will try reconnecting after the final `'close'` (i *think*)

File: app/gui2/shared/languageServer.ts (hunk 3 above)
Comment (somebody1234) [reply app/gui2/shared/languageServer.ts:159]: possibly. i *think* i added this because it logged an error upon closing for some reason, but it is entirely possible that one or more of these guards are no longer needed

File: app/gui2/shared/languageServer.ts (hunk 4 above)
Comment (somebody1234) [reply app/gui2/shared/languageServer.ts]: note that `dispose()` calls `clientScope.dispose()` which in turn calls `transport.close()`

File: app/gui2/shared/languageServer.ts (hunk 3 above)
Comment (somebody1234) [reply app/gui2/shared/languageServer.ts:159]: perhaps instead the event handler should be unregistered using `.off()` like `reinitializeCb`? although i'm not sure it's semantically different than printing things only conditionally

File: app/gui2/shared/languageServer.ts (hunk 4 above)
Comment (farmaazon) [reply app/gui2/shared/languageServer.ts]: > i guess the issue (i think) is that `onAbort` seems to run quite a bit later than `stopReconnecting`, so without this condition then it will try reconnecting after the final `'close'` (i _think_)

I mean, I'm not against this condition, but this line will prevent _initializing_, not reconnecting itself. The reconnecting happens in `transport`'s websocket, regardless of `shouldReconnect` flag here. Therefore, on `close` event, if `shouldReconnect` flag is true, we should manually close transport (which will close websocket manually, which in turn will stop reconnecting attempt).

File: app/gui2/shared/languageServer.ts [hunk 5]
```diff
@@ -159,7 +159,12 @@ export class LanguageServer extends ObservableV2<Notifications & TransportEvents
       }
     })
     const reinitializeCb = () => {
-      if (!this.shouldReconnect) return
+      if (!this.shouldReconnect) {
```
Comment (farmaazon) [inline app/gui2/shared/languageServer.ts]: Please rename `reinitializeCb` to something like `onClosedCb` - it's no longer only about initialization.

**Review by farmaazon — APPROVED** [review]
